src/compiler/checker.ts

```diff
@@ -18350,7 +18356,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     function shouldDeferIndexType(type: Type, indexFlags = IndexFlags.None) {
         return !!(type.flags & TypeFlags.InstantiableNonPrimitive ||
             isGenericTupleType(type) ||
-            isGenericMappedType(type) && (!hasDistributiveNameType(type) || getMappedTypeNameTypeKind(type) === MappedTypeNameTypeKind.Remapping) ||
+            isGenericMappedType(type) && getNameTypeFromMappedType(type) ||

```

Comment:
As mentioned in the comment [here](https://github.com/microsoft/TypeScript/pull/60528/files#r1845690749):
> Effectively P is always instantiated with a non-union type.

Because of that, it's not safe to skip index type deferral for any remapping mapped type as `P` could be used multiple times by the mapping (and the compiler shouldn't allow for a cross-product of `` `${P}_${P}` ``). Checking `getMappedTypeNameTypeKind` can run into an infinite loop here and it seems the easiest to just avoid checking that and assume that all generic mapped types with name types have to have deferred index types. I compensate for that by simplifying them in relationship checking etc.